src/gallium/drivers/tegra/tegra_state.c

```diff
@@ -51,21 +51,49 @@ tegra_set_framebuffer_state(struct pipe_context *pcontext,
 			    const struct pipe_framebuffer_state *framebuffer)
 {
 	struct tegra_context *context = tegra_context(pcontext);
+	struct pipe_framebuffer_state *cso = &context->framebuffer.base;
 	unsigned int i;
+	uint32_t mask = 0;
 
 	for (i = 0; i < framebuffer->nr_cbufs; i++) {
 		struct pipe_surface *ref = framebuffer->cbufs[i];
+		struct tegra_resource *res = tegra_resource(ref->texture);
+		uint32_t rt_params;
+
+		rt_params  = TGR3D_VAL(RT_PARAMS, FORMAT, res->format);
+		rt_params |= TGR3D_VAL(RT_PARAMS, PITCH, res->pitch);
+		rt_params |= TGR3D_BOOL(RT_PARAMS, TILED, res->tiled);
 
-		if (i >= framebuffer->nr_cbufs)
-			ref = NULL;
+		context->framebuffer.rt_params[i] = rt_params;
+		context->framebuffer.bos[i] = res->bo;
+		mask |= 1 << i;
 
-		pipe_surface_reference(&context->framebuffer.base.cbufs[i],
-				       ref);
+		pipe_surface_reference(&cso->cbufs[i], ref);
 	}
 
+	for (; i < cso->nr_cbufs; i++)
+                pipe_surface_reference(&cso->cbufs[i], NULL);
```

Comment:
It's not the first I'm getting confused by the referencing in mesa, could you please remind me how it works? :)

src/gallium/auxiliary/util/u_inlines.h:

 	static inline void
	pipe_surface_reference(struct pipe_surface **ptr, struct pipe_surface *surf)
	{
	struct pipe_surface *old_surf = *ptr;

	if (pipe_reference_described(&(*ptr)->reference, &surf->reference, 
					(debug_reference_descriptor)debug_describe_surface))
	old_surf->context->surface_destroy(old_surf->context, old_surf);
	*ptr = surf;
	}

Wouldn't "surf" be a NULL-deferenced?

And why you are bumping and than decrementing the refcount? Is it some kind of todo here? I'm probably just missing something :)